peas/weather_abstract.py

```diff
@@ -0,0 +1,214 @@
+
+import astropy.units as u
+from astropy.units import cds
+from astropy.table import Table
+from astropy.time import Time, TimeISO, TimeDelta
+
+from pocs.utils.messaging import PanMessaging
+
+from . import load_config
+from .PID import PID
+
+def get_mongodb():
+    from pocs.utils.database import PanMongo
+    return PanMongo()
+
+# -----------------------------------------------------------------------------
+#   Base class to read and check weather data
+# -----------------------------------------------------------------------------
+class WeatherAbstract(object):
+
+    """ Base class for checking generic weather data and sending it to the
+        required location.
+    """
+
+    def __init__(self, use_mongo=True):
+        self.config = load_config()
+
+        self.thresholds = self.config.get('thresholds')
+        self.safety_delay = self.config.get('safety_delay', 15.)
+
+        self.db = None
+        if use_mongo:
+            self.db = get_mongodb()
+
+        self.messaging = None
+        self.safe_dict = None
+        self.weather_entries = list()
+
+    def send_message(self, msg, channel='weather'):
+        if self.messaging is None:
+            self.messaging = PanMessaging.create_publisher(6510)
+
+        self.messaging.send_message(channel, msg)
+
+    def capture(self, data):
+        # Make Safety Decision
+        self.safe_dict = self.make_safety_decision(data)
+
+        data['Safe'] = self.safe_dict['Safe']
+        data['Sky condition'] = self.safe_dict['Sky']
+        data['Wind condition'] = self.safe_dict['Wind']
+        data['Gust condition'] = self.safe_dict['Gust']
+        data['Rain condition'] = self.safe_dict['Rain']
+
+        # Store current weather
+        self.weather_entries.append(data)
+
+        if send_message:
+            self.send_message({'data': data}, channel='weather')
+
+        if use_mongo:
+            self.db.insert_current('weather', data)
+
+        return data
+
+    def make_safety_decision(self, current_values):
+        self.logger.debug('Making safety decision')
+        self.logger.debug('Found {} weather data entries in last {:.0f} minutes'.format(
+            len(self.weather_entries), self.safety_delay))
+        safe = False
+
+        # Tuple with condition,safety
+        cloud = self._get_cloud_safety(current_values)
+
+        try:
+            wind, gust = self._get_wind_safety(current_values)
+        except Exception as e:
+            self.logger.warning('Problem getting wind safety: {}'.format(e))
+            wind = ['N/A']
+            gust = ['N/A']
+
+        rain = self._get_rain_safety(current_values)
+
+        safe = cloud[1] & wind[1] & gust[1] & rain[1]
+        self.logger.debug('Weather Safe: {}'.format(safe))
+
+        return {'Safe': safe,
+                'Sky': cloud[0],
+                'Wind': wind[0],
+                'Gust': gust[0],
+                'Rain': rain[0]}
+
+    def _get_cloud_safety(self, sky_diff, last_cloud):

```

Comment:
Cloud safety should work with the statuses returned from `_get_status()`, not directly with the met data values. That's the point of the `_get_status()` method, it applies all the thresholds all at once in a device independent way.

Reply:
Modified to use the results from `_get_status()`.